<h1>
<center>
Data Science Part 1: Wrangle The Data
</center>
</h1>
<div class=h1_cell>
By Fahad Alarefi. The purpose of this three-part project is to use machine learning into making use of a publicaly available table called IMDB 5000 titles.
The main idea is to come up with a function which can predictate, with high accuracy, whether a provided movie data will have 8 stars or more by the user community. The technique of decision trees will be used.
The dataset is accessed from here:
https://data.world/data-society/imdb-5000-movie-dataset

</div>
<br />
<div class=h1_cell>
In this part, we will wrangle the data.
Wrangling the data means to clean the dataset and make it ready for machine learning!
We will be using Pandas library to clean the data and apply the decision trees.
</div>

<h1>
Read the dataset from a public url
</h1>
<div class=h1_cell>
First step,
get the dataset from the web so that we can work on it.
</div>


In [1]:
import pandas as pd
import os

url = 'https://github.com/createdbyfahad/datascience/files/2593679/movie_metadata.txt'
movie_table = pd.read_csv(url)

pd.set_option('display.max_columns', None)
pd.options.display.float_format = '{:,.0f}'.format

<div class=h1_cell>
Here we can print the top 5 rows from the table:
</div>

In [2]:
movie_table.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                     723       178   
1  Color     Gore Verbinski                     302       169   
2  Color         Sam Mendes                     602       148   
3  Color  Christopher Nolan                     813       164   
4    NaN        Doug Walker                     nan       nan   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                        0                     855  Joel David Moore   
1                      563                   1,000     Orlando Bloom   
2                        0                     161      Rory Kinnear   
3                   22,000                  23,000    Christian Bale   
4                      131                     nan        Rob Walker   

   actor_1_facebook_likes       gross                           genres  \
0                   1,000 760,505,847  Action|Adventure|Fantasy|Sci-Fi   
1                  40,000 309,404,152         Action|Adventure|Fantasy   
2                  11,000 200,074,175        Action|Adventure|Thriller   
3                  27,000 448,130,642                  Action|Thriller   
4                     131         nan                      Documentary   

      actor_1_name                                        movie_title  \
0      CCH Pounder                                            Avatar    
1      Johnny Depp          Pirates of the Caribbean: At World's End    
2  Christoph Waltz                                           Spectre    
3        Tom Hardy                             The Dark Knight Rises    
4      Doug Walker  Star Wars: Episode VII - The Force Awakens    ...   

   num_voted_users  cast_total_facebook_likes          actor_3_name  \
0           886204                       4834             Wes Studi   
1           471220                      48350        Jack Davenport   
2           275868                      11700      Stephanie Sigman   
3          1144337                     106759  Joseph Gordon-Levitt   
4                8                        143                   NaN   

   facenumber_in_poster                                      plot_keywords  \
0                     0             avatar|future|marine|native|paraplegic   
1                     0  goddess|marriage ceremony|marriage proposal|pi...   
2                     1                bomb|espionage|sequel|spy|terrorist   
3                     0  deception|imprisonment|lawlessness|police offi...   
4                     0                                                NaN   

                                     movie_imdb_link  num_user_for_reviews  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...                 3,054   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...                 1,238   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...                   994   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...                 2,701   
4  http://www.imdb.com/title/tt5289954/?ref_=fn_t...                   nan   

  language country content_rating      budget  title_year  \
0  English     USA          PG-13 237,000,000       2,009   
1  English     USA          PG-13 300,000,000       2,007   
2  English      UK          PG-13 245,000,000       2,015   
3  English     USA          PG-13 250,000,000       2,012   
4      NaN     NaN            NaN         nan         nan   

   actor_2_facebook_likes  imdb_score  aspect_ratio  movie_facebook_likes  
0                     936           8             2                 33000  
1                   5,000           7             2                     0  
2                     393           7             2                 85000  
3                  23,000           8             2                164000  
4                      12           7           nan                     0

Check the status of the data (find any empties).


In [3]:
movie_table.isnull().sum()

color                         19
director_name                104
num_critic_for_reviews        50
duration                      15
director_facebook_likes      104
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross                        884
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                153
movie_imdb_link                0
num_user_for_reviews          21
language                      12
country                        5
content_rating               303
budget                       492
title_year                   108
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 329
movie_facebook_likes           0
dtype: int64

Looks like many parts of the data are missing,
however, first we will delete some rows and columns then revaluate empties.
<br />
Since we are interested in modern movies, then first check how many non-color movies are there

In [4]:
len(movie_table.loc[movie_table['color'] != 'Color'])

228

 delete all movies that are not in color

In [5]:
movie_table = movie_table.loc[movie_table['color'] == 'Color']

to make the dataset ligher, we will eliminate all columns that we are not going to use
<br /> 
things like the name of director or the film won't get used in our implementation.

In [6]:
drop_columns = [
     'color',
     'director_name',
     'num_critic_for_reviews',
     'director_facebook_likes',
     'actor_3_facebook_likes',
     'actor_2_name',
     'actor_1_facebook_likes',
     'actor_1_name',
     'movie_title',
     'num_voted_users',
     'cast_total_facebook_likes',
     'actor_3_name',
     'facenumber_in_poster',
     'plot_keywords',
     'movie_imdb_link',
     'num_user_for_reviews',
     'language',
     'actor_2_facebook_likes',
     'aspect_ratio',
     'movie_facebook_likes'
]
movie_table = movie_table.drop(drop_columns, axis=1)
movie_table.head()

duration       gross                           genres country  \
0       178 760,505,847  Action|Adventure|Fantasy|Sci-Fi     USA   
1       169 309,404,152         Action|Adventure|Fantasy     USA   
2       148 200,074,175        Action|Adventure|Thriller      UK   
3       164 448,130,642                  Action|Thriller     USA   
5       132  73,058,679          Action|Adventure|Sci-Fi     USA   

  content_rating      budget  title_year  imdb_score  
0          PG-13 237,000,000       2,009           8  
1          PG-13 300,000,000       2,007           7  
2          PG-13 245,000,000       2,015           7  
3          PG-13 250,000,000       2,012           8  
5          PG-13 263,700,000       2,012           7

now that we have all important information about the each movie, 
<br />
we will start to prepare it for machine learning
<br />
first, check if how many movies have not content_rating

In [7]:
len(movie_table.loc[movie_table['content_rating'].isnull()])

282

since we cannot 'guess' what the movie content type, <br />
then delete all non-rated movies
<br /><br />
also, delete all movies that have not date /OR duration

In [8]:
movie_table = movie_table.loc[movie_table['content_rating'].notnull()]
movie_table = movie_table.loc[movie_table['title_year'].notnull()]
movie_table = movie_table.loc[movie_table['duration'].notnull()]
len(movie_table)

4468

we are only interested in movies, so drop all tv series<br />
if it has "tv" in it's rating, then drop it

In [9]:
movie_table = movie_table[~movie_table['content_rating'].str.contains("TV")]
#movie_table = movie_table[movie_table.apply(lambda row: 'TV' not in row['content_rating'], axis=1)]

we are left with good amount of movie titles
<br />
in order to process the genres, we need to convert them into 1s and 0s
<br >
so we will create a column for each possible genre and fill it with 0 or 1
<br />
It's called one-hot encoding

In [10]:
#movie_table['genres'] = movie_table.genres.str.split('\s*|\s*', expand=True).stack().str.get_dummies().sum(level=0)
movie_table_gd = movie_table['genres'].str.get_dummies(sep='|')
movie_table = movie_table.join(movie_table_gd)
#movie_table += genres
movie_table.head()

duration       gross                           genres country  \
0       178 760,505,847  Action|Adventure|Fantasy|Sci-Fi     USA   
1       169 309,404,152         Action|Adventure|Fantasy     USA   
2       148 200,074,175        Action|Adventure|Thriller      UK   
3       164 448,130,642                  Action|Thriller     USA   
5       132  73,058,679          Action|Adventure|Sci-Fi     USA   

  content_rating      budget  title_year  imdb_score  Action  Adventure  \
0          PG-13 237,000,000       2,009           8       1          1   
1          PG-13 300,000,000       2,007           7       1          1   
2          PG-13 245,000,000       2,015           7       1          1   
3          PG-13 250,000,000       2,012           8       1          0   
5          PG-13 263,700,000       2,012           7       1          1   

   Animation  Biography  Comedy  Crime  Documentary  Drama  Family  Fantasy  \
0          0          0       0      0            0      0       0        1   
1          0          0       0      0            0      0       0        1   
2          0          0       0      0            0      0       0        0   
3          0          0       0      0            0      0       0        0   
5          0          0       0      0            0      0       0        0   

   Film-Noir  History  Horror  Music  Musical  Mystery  News  Romance  Sci-Fi  \
0          0        0       0      0        0        0     0        0       1   
1          0        0       0      0        0        0     0        0       0   
2          0        0       0      0        0        0     0        0       0   
3          0        0       0      0        0        0     0        0       0   
5          0        0       0      0        0        0     0        0       1   

   Sport  Thriller  War  Western  
0      0         0    0        0  
1      0         0    0        0  
2      0         1    0        0  
3      0         1    0        0  
5      0         0    0        0

drop movies with nan country, <br />
and then one hot encode the coutnry

In [11]:
movie_table = movie_table.loc[movie_table['country'].notnull()]
one_hot_c = pd.get_dummies(movie_table['country'],prefix='c',dummy_na=False)
movie_table = movie_table.join(one_hot_c)
movie_table.head()

duration       gross                           genres country  \
0       178 760,505,847  Action|Adventure|Fantasy|Sci-Fi     USA   
1       169 309,404,152         Action|Adventure|Fantasy     USA   
2       148 200,074,175        Action|Adventure|Thriller      UK   
3       164 448,130,642                  Action|Thriller     USA   
5       132  73,058,679          Action|Adventure|Sci-Fi     USA   

  content_rating      budget  title_year  imdb_score  Action  Adventure  \
0          PG-13 237,000,000       2,009           8       1          1   
1          PG-13 300,000,000       2,007           7       1          1   
2          PG-13 245,000,000       2,015           7       1          1   
3          PG-13 250,000,000       2,012           8       1          0   
5          PG-13 263,700,000       2,012           7       1          1   

   Animation  Biography  Comedy  Crime  Documentary  Drama  Family  Fantasy  \
0          0          0       0      0            0      0       0        1   
1          0          0       0      0            0      0       0        1   
2          0          0       0      0            0      0       0        0   
3          0          0       0      0            0      0       0        0   
5          0          0       0      0            0      0       0        0   

   Film-Noir  History  Horror  Music  Musical  Mystery  News  Romance  Sci-Fi  \
0          0        0       0      0        0        0     0        0       1   
1          0        0       0      0        0        0     0        0       0   
2          0        0       0      0        0        0     0        0       0   
3          0        0       0      0        0        0     0        0       0   
5          0        0       0      0        0        0     0        0       1   

   Sport  Thriller  War  Western  c_Afghanistan  c_Argentina  c_Aruba  \
0      0         0    0        0              0            0        0   
1      0         0    0        0              0            0        0   
2      0         1    0        0              0            0        0   
3      0         1    0        0              0            0        0   
5      0         0    0        0              0            0        0   

   c_Australia  c_Bahamas  c_Belgium  c_Brazil  c_Bulgaria  c_Cameroon  \
0            0          0          0         0           0           0   
1            0          0          0         0           0           0   
2            0          0          0         0           0           0   
3            0          0          0         0           0           0   
5            0          0          0         0           0           0   

   c_Canada  c_Chile  c_China  c_Colombia  c_Czech Republic  c_Denmark  \
0         0        0        0           0                 0          0   
1         0        0        0           0                 0          0   
2         0        0        0           0                 0          0   
3         0        0        0           0                 0          0   
5         0        0        0           0                 0          0   

   c_Dominican Republic  c_Egypt  c_Finland  c_France  c_Georgia  c_Germany  \
0                     0        0          0         0          0          0   
1                     0        0          0         0          0          0   
2                     0        0          0         0          0          0   
3                     0        0          0         0          0          0   
5                     0        0          0         0          0          0   

   c_Greece  c_Hong Kong  c_Hungary  c_Iceland  c_India  c_Indonesia  c_Iran  \
0         0            0          0          0        0            0       0   
1         0            0          0          0        0            0       0   
2         0            0          0          0        0            0       0   
3         0            0          0          0        0  

similarly, one hot encode the content_rating <br />
also, I have noticed that some movies use the old rating system (M, GP, G, and X).
<br /> so I will change it to the newer systm

In [12]:
one_hot_r = pd.get_dummies(movie_table['content_rating'],prefix='rating',dummy_na=False)
movie_table = movie_table.join(one_hot_r)
#movie_table['rating_PG'] = movie_table.apply(lambda row: row['rating_G'])
movie_table['rating_PG'] = (movie_table['rating_G'] | movie_table['rating_PG'])
movie_table['rating_PG-13'] = (movie_table['rating_GP'] | movie_table['rating_M'] | movie_table['rating_PG-13'])
movie_table['rating_NC-17'] = (movie_table['rating_X'] | movie_table['rating_NC-17'])
movie_table['rating_Unrated'] = (movie_table['rating_Not Rated'] | movie_table['rating_Approved'] | movie_table['rating_Unrated'] | movie_table['rating_Passed'])
# then delete unnessasery columns
movie_table = movie_table.drop(['rating_G', 'rating_M', 'rating_GP', 'rating_X', 'rating_Not Rated', 'rating_Approved', 'rating_Passed'], axis=1)
movie_table.head()

duration       gross                           genres country  \
0       178 760,505,847  Action|Adventure|Fantasy|Sci-Fi     USA   
1       169 309,404,152         Action|Adventure|Fantasy     USA   
2       148 200,074,175        Action|Adventure|Thriller      UK   
3       164 448,130,642                  Action|Thriller     USA   
5       132  73,058,679          Action|Adventure|Sci-Fi     USA   

  content_rating      budget  title_year  imdb_score  Action  Adventure  \
0          PG-13 237,000,000       2,009           8       1          1   
1          PG-13 300,000,000       2,007           7       1          1   
2          PG-13 245,000,000       2,015           7       1          1   
3          PG-13 250,000,000       2,012           8       1          0   
5          PG-13 263,700,000       2,012           7       1          1   

   Animation  Biography  Comedy  Crime  Documentary  Drama  Family  Fantasy  \
0          0          0       0      0            0      0       0        1   
1          0          0       0      0            0      0       0        1   
2          0          0       0      0            0      0       0        0   
3          0          0       0      0            0      0       0        0   
5          0          0       0      0            0      0       0        0   

   Film-Noir  History  Horror  Music  Musical  Mystery  News  Romance  Sci-Fi  \
0          0        0       0      0        0        0     0        0       1   
1          0        0       0      0        0        0     0        0       0   
2          0        0       0      0        0        0     0        0       0   
3          0        0       0      0        0        0     0        0       0   
5          0        0       0      0        0        0     0        0       1   

   Sport  Thriller  War  Western  c_Afghanistan  c_Argentina  c_Aruba  \
0      0         0    0        0              0            0        0   
1      0         0    0        0              0            0        0   
2      0         1    0        0              0            0        0   
3      0         1    0        0              0            0        0   
5      0         0    0        0              0            0        0   

   c_Australia  c_Bahamas  c_Belgium  c_Brazil  c_Bulgaria  c_Cameroon  \
0            0          0          0         0           0           0   
1            0          0          0         0           0           0   
2            0          0          0         0           0           0   
3            0          0          0         0           0           0   
5            0          0          0         0           0           0   

   c_Canada  c_Chile  c_China  c_Colombia  c_Czech Republic  c_Denmark  \
0         0        0        0           0                 0          0   
1         0        0        0           0                 0          0   
2         0        0        0           0                 0          0   
3         0        0        0           0                 0          0   
5         0        0        0           0                 0          0   

   c_Dominican Republic  c_Egypt  c_Finland  c_France  c_Georgia  c_Germany  \
0                     0        0          0         0          0          0   
1                     0        0          0         0          0          0   
2                     0        0          0         0          0          0   
3                     0        0          0         0          0          0   
5                     0        0          0         0          0          0   

   c_Greece  c_Hong Kong  c_Hungary  c_Iceland  c_India  c_Indonesia  c_Iran  \
0         0            0          0          0        0            0       0   
1         0            0          0          0        0            0       0   
2         0            0          0          0        0            0       0   
3         0            0          0          0        0  

Now we are done with content rating, <br />
next, we will classify the numbers data, like 
new_title = if title >= 2000
length_avg = if duration > average movie duration (130 minutes)
good_score = if imdb_score >= 8

In [13]:
movie_table['is_new'] = movie_table.apply(lambda row: 1 if row.title_year >= 2000.0 else 0, axis=1)
movie_table['is_long'] = movie_table.apply(lambda row: 1 if row.duration > 130.0 else 0, axis=1)
movie_table['is_good'] = movie_table.apply(lambda row: 1 if row.imdb_score >= 8.0 else 0, axis=1)

movie_table.head()

duration       gross                           genres country  \
0       178 760,505,847  Action|Adventure|Fantasy|Sci-Fi     USA   
1       169 309,404,152         Action|Adventure|Fantasy     USA   
2       148 200,074,175        Action|Adventure|Thriller      UK   
3       164 448,130,642                  Action|Thriller     USA   
5       132  73,058,679          Action|Adventure|Sci-Fi     USA   

  content_rating      budget  title_year  imdb_score  Action  Adventure  \
0          PG-13 237,000,000       2,009           8       1          1   
1          PG-13 300,000,000       2,007           7       1          1   
2          PG-13 245,000,000       2,015           7       1          1   
3          PG-13 250,000,000       2,012           8       1          0   
5          PG-13 263,700,000       2,012           7       1          1   

   Animation  Biography  Comedy  Crime  Documentary  Drama  Family  Fantasy  \
0          0          0       0      0            0      0       0        1   
1          0          0       0      0            0      0       0        1   
2          0          0       0      0            0      0       0        0   
3          0          0       0      0            0      0       0        0   
5          0          0       0      0            0      0       0        0   

   Film-Noir  History  Horror  Music  Musical  Mystery  News  Romance  Sci-Fi  \
0          0        0       0      0        0        0     0        0       1   
1          0        0       0      0        0        0     0        0       0   
2          0        0       0      0        0        0     0        0       0   
3          0        0       0      0        0        0     0        0       0   
5          0        0       0      0        0        0     0        0       1   

   Sport  Thriller  War  Western  c_Afghanistan  c_Argentina  c_Aruba  \
0      0         0    0        0              0            0        0   
1      0         0    0        0              0            0        0   
2      0         1    0        0              0            0        0   
3      0         1    0        0              0            0        0   
5      0         0    0        0              0            0        0   

   c_Australia  c_Bahamas  c_Belgium  c_Brazil  c_Bulgaria  c_Cameroon  \
0            0          0          0         0           0           0   
1            0          0          0         0           0           0   
2            0          0          0         0           0           0   
3            0          0          0         0           0           0   
5            0          0          0         0           0           0   

   c_Canada  c_Chile  c_China  c_Colombia  c_Czech Republic  c_Denmark  \
0         0        0        0           0                 0          0   
1         0        0        0           0                 0          0   
2         0        0        0           0                 0          0   
3         0        0        0           0                 0          0   
5         0        0        0           0                 0          0   

   c_Dominican Republic  c_Egypt  c_Finland  c_France  c_Georgia  c_Germany  \
0                     0        0          0         0          0          0   
1                     0        0          0         0          0          0   
2                     0        0          0         0          0          0   
3                     0        0          0         0          0          0   
5                     0        0          0         0          0          0   

   c_Greece  c_Hong Kong  c_Hungary  c_Iceland  c_India  c_Indonesia  c_Iran  \
0         0            0          0          0        0            0       0   
1         0            0          0          0        0            0       0   
2         0            0          0          0        0            0       0   
3         0            0          0          0        0  

Here are some important functions that will help in applying the kmeans values
(source: Stephen Ficaks, University of Oregon)

In [14]:
def row_to_vect(row, features):
    vect = []
    for feature in features:
        vect.append(float(row[feature]))
    return tuple(vect)

def initialize_centroids(sample_table, features):
    k = len(sample_table)
    centroids = {}
    for i in range(k):
        row = sample_table.iloc[i]
        vector = row_to_vect(row, features)
        centroids[i] = {'centroid': vector, 'cluster': []}
        
    return centroids

def euclidean_distance(vect1, vect2):
    sum = 0
    for i in range(len(vect1)):
        sum += (vect1[i] - vect2[i])**2
    return sum**.5  # I claim that this square root is not needed in K-means - see why?

def closest_centroid(centroids, row, k):
    min_distance = euclidean_distance(centroids[0]['centroid'], row)
    min_centroid = 0
    for i in range(1,k):
        distance = euclidean_distance(centroids[i]['centroid'], row)
        if distance < min_distance:
            min_distance = distance
            min_centroid = i
    return (min_centroid, min_distance)

def phase_1(centroids, table, features, k):
    for i in range(k):
        centroids[i]['cluster'] = []  # starting new phase 1 so empty out values from prior iteration
    
    #Go through every row in Titanic table (or Loan table) and place in closest centroid cluster.
    for i in range(len(table)):
        row = table.iloc[i]
        vrow = row_to_vect(row, features)
        (index, dist) = closest_centroid(centroids, vrow, k)
        centroids[index]['cluster'].append(vrow)
    
    return centroids

#cluster is a list of points, i.e., a list of lists.
def compute_mean(cluster):
    if len(cluster) == 0:
        return []
    the_sum = cluster[0]  # use 0th point as starter
    
    #I am using zip to pair up all points then do addition
    for i in range(1,len(cluster)):
        the_sum = [pair[0]+pair[1] for pair in zip(the_sum, cluster[i])]
    n = len(cluster)*1.0
    the_mean_point = [x/n for x in the_sum]
    return the_mean_point


def phase_2(centroids, k, threshold):
    old_centroids = []
    
    stop = True
    #Compute k new centroids and check for stopping condition
    for i in range(k):
        current_centroid = centroids[i]['centroid']
        new_centroid = compute_mean(centroids[i]['cluster'])
        centroids[i]['centroid'] = new_centroid
        if euclidean_distance(current_centroid, new_centroid) > threshold:
            stop = False  # all it takes is one

    return (stop, centroids)


def k_means(table, features, k, hypers):
    n = 100 if 'n' not in hypers else hypers['n']
    threshold = 0.0 if 'threshold' not in hypers else hypers['threshold']
    
    centroid_table = table.sample(n=k, replace=False, random_state=100)  # only random choice I am making
    centroids = initialize_centroids(centroid_table, features)
    
    j = 0
    stop = False
    while( j < n and not stop):
        print('starting '+str(j+1))
        centroids = phase_1(centroids, table, features, k)
        (stop, centroids) = phase_2(centroids, k, threshold)
        j += 1
    print('done')
    return centroids


In [15]:
len(movie_table.loc[movie_table['gross'].isnull()])

502

gross is not so important field, but it will help us in figuring out the missing budget values
so as you see above, gross has a lot empty values
<br /> so using kmeans, we will fill it with values


In [16]:
len(movie_table.loc[movie_table['duration'].isnull()])

0

In [17]:
features_used = [u'duration',
        u'title_year', u'imdb_score', u'Action', u'Adventure',
       u'Animation', u'Biography', u'Comedy', u'Crime', u'Documentary',
       u'Drama', u'Family', u'Fantasy', u'Film-Noir', u'History', u'Horror',
       u'Music', u'Musical', u'Mystery', u'News', u'Romance', u'Sci-Fi',
       u'Sport', u'Thriller', u'War', u'Western', u'c_Afghanistan',
       u'c_Argentina', u'c_Aruba', u'c_Australia', u'c_Bahamas', u'c_Belgium',
       u'c_Brazil', u'c_Bulgaria', u'c_Cameroon', u'c_Canada', u'c_Chile',
       u'c_China', u'c_Colombia', u'c_Czech Republic', u'c_Denmark',
       u'c_Dominican Republic', u'c_Egypt', u'c_Finland', u'c_France',
       u'c_Georgia', u'c_Germany', u'c_Greece', u'c_Hong Kong', u'c_Hungary',
       u'c_Iceland', u'c_India', u'c_Indonesia', u'c_Iran', u'c_Ireland',
       u'c_Israel', u'c_Italy', u'c_Japan', u'c_Kyrgyzstan', u'c_Mexico',
       u'c_Netherlands', u'c_New Line', u'c_New Zealand', u'c_Norway',
       u'c_Official site', u'c_Panama', u'c_Peru', u'c_Philippines',
       u'c_Poland', u'c_Romania', u'c_Russia', u'c_Slovakia',
       u'c_South Africa', u'c_South Korea', u'c_Spain', u'c_Sweden',
       u'c_Taiwan', u'c_Thailand', u'c_UK', u'c_USA', u'c_West Germany',
       u'rating_NC-17', u'rating_PG', u'rating_PG-13', u'rating_R',
       u'rating_Unrated']


In [18]:
k = 5
final_centroids = k_means(movie_table, features_used, k, {'n':100, 'threshold':0.0})

starting 1
starting 2
starting 3
starting 4
starting 5
starting 6
starting 7
starting 8
starting 9
starting 10
starting 11
starting 12
starting 13
starting 14
starting 15
starting 16
starting 17
starting 18
starting 19
starting 20
starting 21
starting 22
starting 23
starting 24
starting 25
starting 26
starting 27
starting 28
starting 29
starting 30
starting 31
starting 32
done


now we need to create a table that has all rows with nonempty 'gross' value
we will use that table to get a guess the empty value in our table

In [19]:
focus_col = 'gross'

nonempty_gross_table = movie_table.loc[movie_table[focus_col].notnull()]
print(len(nonempty_gross_table))

3956


In [20]:
def compute_centroid_labels(centroids, focus_table, focus_column, features, k):
    for i in range(len(focus_table)):
        row = focus_table.iloc[i]
        vrow = row_to_vect(row, features)
        (minc, mind) = closest_centroid(centroids, vrow, k)
        if focus_column not in centroids[minc]:
            centroids[minc][focus_column] = [row[focus_column]*1.0]
        else:
            centroids[minc][focus_column].append(row[focus_column]*1.0)
    for ind in range(k): 
        if len(centroids[ind][focus_column]) == 0:
            centroids[ind]['mean_label'] = 0.0
        else:
            the_sum = centroids[ind][focus_column][0]
            for i in range(1, len(centroids[ind][focus_column])):
                the_sum += centroids[ind][focus_column][i]
            centroids[ind]['mean_label'] = the_sum/(len(centroids[ind][focus_column]) * 1.0)
            
    return centroids
    
centroids = compute_centroid_labels(final_centroids, nonempty_gross_table, focus_col, features_used, k)

now we will fill the empty 'gross' values in the table.


In [21]:
def kmeans_fill(centroids, full_table, features, focus_column, k):
   
    def get_closest_clmn(row):
        vrow = row_to_vect(row, features)
        (minc, mind) = closest_centroid(centroids, vrow, k)
        
        return centroids[minc]['mean_label']
    
    new_table = pd.DataFrame(full_table)
    new_table['kmeans_'+focus_column] = new_table.apply(lambda row: get_closest_clmn(row) if pd.isnull(row[focus_column]) else row[focus_column], axis=1)
    
    return new_table

movie_table = kmeans_fill(centroids, movie_table, features_used, focus_col, k)
movie_table.loc[movie_table[focus_col].isnull()].head(20)

duration  gross                                            genres  \
84         109    nan                   Action|Adventure|Romance|Sci-Fi   
367        141    nan                           Adventure|Drama|History   
422        123    nan                        Comedy|Drama|Music|Musical   
430        130    nan                             Comedy|Fantasy|Horror   
507         89    nan                                  Animation|Family   
676        114    nan                                  Action|Drama|War   
891         96    nan                     Crime|Horror|Mystery|Thriller   
1060        93    nan                            Action|Adventure|Drama   
1072       140    nan                                 Drama|History|War   
1099       111    nan                                     Comedy|Family   
1136        91    nan            Action|Animation|Comedy|Family|Fantasy   
1326        90    nan  Adventure|Animation|Comedy|Fantasy|Music|Romance   
1365        97    nan                  Adventure|Comedy|History|Romance   
1446       102    nan                         Comedy|Drama|Family|Sport   
1459       111    nan                           Action|Adventure|Sci-Fi   
1463        93    nan                             Action|Crime|Thriller   
1481        93    nan                              Action|Horror|Sci-Fi   
1482       116    nan                          Action|Adventure|Fantasy   
1508       109    nan                   Action|Adventure|Romance|Sci-Fi   
1509        93    nan                       Action|Comedy|Crime|Western   

          country content_rating     budget  title_year  imdb_score  Action  \
84        Belgium              R        nan       2,015           4       1   
367           USA          PG-13        nan       2,016           6       0   
422           USA              R        nan       1,979           8       0   
430           USA              R  8,000,000       1,982           7       0   
507        Mexico             PG  8,000,000       2,015           4       0   
676           USA          PG-13 16,900,000       1,984           6       1   
891           USA              R 55,000,000       2,002           5       0   
1060          USA              R 14,000,000       1,982           8       1   
1072  South Korea             PG 48,000,000       1,981           3       0   
1099          USA        Unrated  2,500,000       1,968           7       0   
1136          USA             PG 65,000,000       2,012           2       1   
1326       France             PG 25,000,000       2,011           7       0   
1365        Italy              R 38,000,000       2,007           5       0   
1446          USA             PG        nan       1,976           7       0   
1459           UK             PG 20,000,000       1,980           6       1   
1463          USA              R 19,000,000       2,015           6       1   
1481    Australia              R 30,000,000       2,012           5       1   
1482           UK             PG 27,000,000       1,983           6       1   
1508      Belgium              R        nan       2,015           4       1   
1509       France          PG-13 32,000,000       2,006           6       1   

      Adventure  Animation  Biography  Comedy  Crime  Documentary  Drama  \
84            1          0          0       0      0            0      0   
367           1          0          0       0      0            0      1   
422           0          0          0       1      0            0      1   
430           0          0          0       1      0            0      0   
507           0          1          0       0      0            0      0   
676           0          0          0       0      0            0      1   
891           0          0          0       0      1            0      0   
1060          1          0          0       0      0            0      1   
1072          0          0          0       0      0            0      1   
1099      

As you see above, all empty 'gross' values have kmeans value in the same row
<br /> <br />
now we will do the same thing for the budget column and get a kmeans estimate
<br /> 
but since we have a 'gross' value for every movie, then we will use that info <br />
to gain better estimate for the budget.

In [22]:
features_used.append('kmeans_gross')

we will redo what we last did, but now try to get the budget value

In [23]:
focus_col = 'budget'

nonempty_budget_table = movie_table.loc[movie_table[focus_col].notnull()]
print(len(nonempty_budget_table))

4141


In [24]:
second_final_centroids = k_means(movie_table, features_used, k, {'n':100, 'threshold':0.0})
second_centroids = compute_centroid_labels(second_final_centroids, nonempty_budget_table, focus_col, features_used, k)

starting 1
starting 2
starting 3
starting 4
starting 5
starting 6
starting 7
starting 8
starting 9
starting 10
starting 11
starting 12
starting 13
starting 14
starting 15
starting 16
starting 17
starting 18
starting 19
starting 20
starting 21
starting 22
starting 23
starting 24
starting 25
starting 26
starting 27
starting 28
starting 29
starting 30
done


In [25]:
movie_table = kmeans_fill(second_centroids, movie_table, features_used, focus_col, k)
movie_table.loc[movie_table[focus_col].isnull()].head(20)

duration       gross                                     genres  \
55          93 123,070,338  Adventure|Animation|Comedy|Family|Fantasy   
84         109         nan            Action|Adventure|Romance|Sci-Fi   
367        141         nan                    Adventure|Drama|History   
422        123         nan                 Comedy|Drama|Music|Musical   
444         97  82,506,325                      Comedy|Family|Fantasy   
477        117     821,997                              Drama|Romance   
490        117  27,779,888            Action|Adventure|Drama|Thriller   
559         84      90,820               Drama|Horror|Sci-Fi|Thriller   
640        117  27,550,735              Action|Drama|History|Thriller   
642        114  72,980,108                                     Comedy   
713        100 168,213,584                    Action|Adventure|Comedy   
773        127  36,965,395               Crime|Drama|Mystery|Thriller   
780        111  35,754,555                                Drama|Sport   
789         86  28,424,210            Animation|Comedy|Family|Fantasy   
925         93  35,092,918                              Comedy|Family   
949        124  75,621,915                        Comedy|Drama|Family   
952        128  75,888,270                             Horror|Mystery   
1037       135   7,994,115                         Comedy|Drama|Music   
1040       129  12,802,907                              Drama|Romance   
1056       110     528,972                          Drama|Romance|War   

      country content_rating  budget  title_year  imdb_score  Action  \
55        USA             PG     nan       2,015           7       0   
84    Belgium              R     nan       2,015           4       1   
367       USA          PG-13     nan       2,016           6       0   
422       USA              R     nan       1,979           8       0   
444       USA              G     nan       2,006           6       0   
477       USA          PG-13     nan       1,997           7       0   
490       USA              R     nan       1,997           7       1   
559       USA          PG-13     nan       2,009           6       0   
640       USA          PG-13     nan       2,016           7       1   
642       USA          PG-13     nan       2,010           6       0   
713       USA          PG-13     nan       2,007           6       1   
773       USA          PG-13     nan       2,009           7       0   
780       USA          PG-13     nan       2,012           7       0   
789        UK             PG     nan       2,006           5       0   
925       USA             PG     nan       2,006           5       0   
949       USA             PG     nan       2,011           7       0   
952       USA          PG-13     nan       2,005           5       0   
1037       UK              R     nan       2,009           7       0   
1040       UK              R     nan       2,012           7       0   
1056    India        Unrated     nan       1,998           8       0   

      Adventure  Animation  Biography  Comedy  Crime  Documentary  Drama  \
55            1          1          0       1      0            0      0   
84            1          0          0       0      0            0      0   
367           1          0          0       0      0            0      1   
422           0          0          0       1      0            0      1   
444           0          0          0       1      0            0      0   
477           0          0          0       0      0            0      1   
490           1          0          0       0      0            0      1   
559           0          0          0       0      0            0      1   
640           0          0          0       0      0            0      1   
642           0          0          0       1      0            0      0   
713           1          0          0       1      0            0      0   
773           0          0          0       0  

Now, we are going to use this data to further examination. <br />
In the the next two parts we will try two models of prediciton. (Random Forests and Decision Trees) <br />
In order to use reuse this data, we need to store in a file.


In [26]:
import os

movie_table.to_csv("part1_movie_table.csv", index=False)